examples/demo/src/controllers/middleware/api_key.rs

```diff
@@ -0,0 +1,42 @@
+//! Axum middleware for validating API key tokens.
+
+use async_trait::async_trait;
+use axum::{
+    extract::{FromRef, FromRequestParts},
+    http::request::Parts,
+};
+use loco_rs::{app::AppContext, controller::middleware::auth, errors::Error};
+use serde::{Deserialize, Serialize};
+
+use crate::models::_entities::users;
+
+#[derive(Debug, Deserialize, Serialize)]
+// Represents the data structure for the API token.
+pub struct ApiToken {
```

Comment:
I am wondering,
how possible is it to say:

```
pub struct ApiToken<T>
 ... everything depends on T
 ... includng the extractor 
```

and we know that T should eventually be users::Model,
maybe this T will be a `T: Trait` where `T: FetchByApiKey`,

and the `User impl` that FetchByApiKey.

Now this middleware becomes shareable by everyone

Reply:
The API token is generated within the user app (Loco framework is not familiar with that). In this scenario, I think the <T> parameter is meaningless, as the user has complete control over it.

About the FetchByJWT i totally agree. maybe we can consider moving the claims to be a <T>

Reply:
Notice what I did in the latest commit. I had a feeling if we move things around we might get something, I think it worked:

`Authenticable` trait lets us encapsulate all the logic, and then -- just refer to an extractor. Use `ApiKey<T>` and `ApiKey<users::Model>` to give Rust all the type information it needs to resolve.

On this "Authenticable" concept we can also encode the `JWT` behavior, and "capture" a user just like in api key auth.
There is a good chance that instead of having _claims_ we can have the actual user model if we use the same trick.